File: LeetCode/144二叉树的前序遍历.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */


/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
void pre_order(struct TreeNode *root, int *res, int *returnSize) {
    if (root == NULL) return ;
    res[(*returnSize)++] = root->val;
    pre_order(root->left, res, returnSize);
    pre_order(root->right, res, returnSize);
    return ;
}

int* preorderTraversal(struct TreeNode* root, int* returnSize){
    int *res = (int *)malloc(sizeof(int) * 505);
    *returnSize = 0;
    pre_order(root, res, res, returnSize);
    return res;
}

File: LeetCode/144二叉树的前序遍历.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void preOrder(TreeNode root, List<Integer> ans) {
        if (root == null) return ;
        ans.add(root.val);
        preOrder(root.left, ans);
        preOrder(root.right, ans);
    }

    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        preOrder(root, ans);
        return ans;
    }
}

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        Stack<TreeNode> sta = new Stack<>();
        if (root != null) sta.push(root);
        while (!sta.isEmpty()) {
            TreeNode cur = sta.pop();
            ans.add(cur.val);
            if (cur.right != null) sta.push(cur.right);
            if (cur.left != null) sta.push(cur.left);
        }
        return ans;
    }
}